core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableJoinRule.java

```diff
@@ -61,12 +61,29 @@
     if (!info.isEqui() && join.getJoinType() != JoinRelType.INNER) {
```

Comment:
Is this `if` condition still relevant? The scenario of a non-equi inner join implemented as an EnumerableHashJoin + Filter should be supported with just a (new version of) EnumerableHashJoin, which can take the full condition.
I have the impression that we can remove the  `if` condition (and the whole else block) and leave just the if block: if there are equiKeys, generate EnumerableHashJoin (which now supports equi and non-equi conditions); otherwise generate EnumerableNestedLoopJoin.

Reply:
yes, since the SEMI/ANTI join that with empty nonEquiConditions is handled by EnumerableHashJoin, the `if` condition should handle those case.